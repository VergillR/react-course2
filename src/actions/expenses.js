// import uuid from 'uuid'
import db from '../firebase/firebase'

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '', note = '', amount = 0, createdAt = 0
    } = expenseData
    const newExpense = { description, note, amount, createdAt }

    return db.ref('expenses').push(newExpense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...newExpense
      }))
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id = '' } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

// EDIT_EXPENSE
export const editExpense = (id = '', updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// export const startSetExpenses
// 1. fetch all expense data now
// 2. parse that data into an array
// 3. dispatch SET_EXPENSES
export const startSetExpenses = () => {
  return (dispatch) => {
    return db.ref('expenses').once('value', (snapshot) => {
      const temp = []
      snapshot.forEach((childSnapshot) => {
        temp.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setExpenses(temp))
    })
  }
}

export const startRemoveExpense = (exp = {}) => {
  return (dispatch) => {
    return db.ref(`expenses/${exp.id}`).remove().then(() => {
      dispatch(removeExpense(exp))
    }).catch((err) => {
      console.log(err)
    })
  }
}
