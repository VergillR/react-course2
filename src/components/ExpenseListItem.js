// Export a stateless functional component
// description, amount, createdAt
import React from 'react'
import { Link } from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'

// load a locale
numeral.register('locale', 'nl', {
  delimiters: {
    thousands: '.',
    decimal: ','
  },
  abbreviations: {
    thousand: 'k',
    million: 'mln',
    billion: 'mrd',
    trillion: 'bln'
  },
  ordinal: function (number) {
    var remainder = number % 100
    return (number !== 0 && remainder <= 1 || remainder === 8 || remainder >= 20) ? 'ste' : 'de'
  },
  currency: {
    symbol: 'â‚¬ '
  }
})

// switch between locales
numeral.locale('nl')

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <h2>Expense</h2>
    <Link to={'/edit/' + id}><h3>Description: {description}</h3></Link>
    <p>Amount: {numeral(amount / 100).format('$0,0.00')}</p>
    <p>CreatedAt: {moment(createdAt).format('MMMM Do, YYYY')}</p>
  </div>
)

export default ExpenseListItem
