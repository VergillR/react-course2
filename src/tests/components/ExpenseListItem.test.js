/* global test, expect */
import React from 'react'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'
import ExpenseListItem from '../../components/ExpenseListItem'

test('should render expense list item with expenses[0] data', () => {
  const exp = expenses[0]
  const wrapper = shallow(<ExpenseListItem {...exp} />)
  expect(wrapper).toMatchSnapshot()
})

