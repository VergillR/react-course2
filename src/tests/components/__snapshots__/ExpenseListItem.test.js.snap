// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item with expenses[0] data 1`] = `
<div>
  <h2>
    Expense
  </h2>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Description: 
      Coffee
    </h3>
  </Link>
  <p>
    Amount: 
    â‚¬ 18,00
  </p>
  <p>
    CreatedAt: 
    January 1st, 1970
  </p>
</div>
`;
